pkg/kubelet/nodelease/controller.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodelease
+
+import (
+	"time"
+
+	coordv1beta1 "k8s.io/api/coordination/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	coordclientset "k8s.io/client-go/kubernetes/typed/coordination/v1beta1"
+	"k8s.io/kubernetes/pkg/util/pointer"
+
+	"github.com/golang/glog"
+)
+
+const (
+	// renewFactor is the fraction of the lease duration that the Kubelet waits before renewing the lease
+	// TODO(mtaufen): determine a good renewFactor, this should be a standard behavior and should not need to be configured
+	renewFactor = 0.33
+	// maxUpdateRetries is the number of immediate, successive retries the Kubelet will attempt
+	// when renewing the lease before it waits for the renewal interval before trying again,
+	// similar to what we do for node status retries
+	maxUpdateRetries = 5
+	// maxBackoff is the maximum sleep time during backoff (e.g. in backoffEnsureLease)
+	maxBackoff = 7 * time.Second
+)
+
+// Controller manages creating and renewing the lease for this Kubelet
+type Controller interface {
+	Run(stopCh <-chan struct{})
+}
+
+var _ Controller = &controller{}
+
+type controller struct {
+	client                     coordclientset.LeaseInterface
+	holderIdentity             string
+	leaseDurationSeconds       int32
+	renewInterval              time.Duration
+	clock                      clock.Clock
+	onRepeatedHeartbeatFailure func()
+}
+
+// NewController constructs and returns a controller
+func NewController(clock clock.Clock, client clientset.Interface, holderIdentity string, leaseDurationSeconds int32, onRepeatedHeartbeatFailure func()) Controller {
+	var leaseClient coordclientset.LeaseInterface
+	if client != nil {
+		leaseClient = client.CoordinationV1beta1().Leases(corev1.NamespaceNodeLease)
+	}
+	return &controller{
+		client:               leaseClient,
+		holderIdentity:       holderIdentity,
+		leaseDurationSeconds: leaseDurationSeconds,
+		renewInterval:        time.Duration(float64(time.Duration(leaseDurationSeconds)*time.Second) * renewFactor),
```

Comment:
I want to start thinking about configuration in terms of higher-level policy controls, which will help us manage configuration growth over the next several years.

IMO the things we care about here are:
- How long the node lifecycle controller waits to cordon an unhealthy node (lease duration).
- How tolerant the node lifecycle controller's decisions are to network disruptions, etc. 

The former is a higher-level policy control, and I'd expect it to eventually move to some in-cluster policy object that both Kubelet and NLC read, rather than being plumbed through low-level config.

The latter is a function of the update frequency and lease duration; with a longer lease, you don't need to update as frequently to have the same tolerance. I think we can settle on a standard, acceptable principle for tolerance, and not have to grow our configuration space.

Reply:
I think it's likely that folks would request to configure the renewal interval further in the future. If that happens, exposing "renew interval" (or something similar) externally is better exposing the `renewFactor` IMO. The interpretation of the latter may depend on the internal logic you come up with and is subject to change. There could also be corner cases such as setting the minimal realistic interval, etc.

Besides that (which is not yet a problem), I think the in the KEP we agreed to start with the 10s heartbeat (i.e., lease renewal interval) by default so that external controllers/components relying on the the existing update frequency can migrate to watch the leases instead.